### DISCUSSION TITLE
Proposal: attributes as first class language members ("has" keyword)

### DISCUSSION THREAD

**Discussion by pelletburner** [dotnet/csharplang#6733] [Language Ideas] [answered]
This relates to polymorphism and I'd like to use something like this:

```
public abstract class SomethingBase
{
  // whatever
}

public sealed class TheRealThing : SomethingBase
{
  // whatever
}

[Unreal]
public sealed class The UnrealThing : SomethingBase
{
  // whatever
}
```

Using it now:

```
SomethingBase real = new TheRealThing();
SomethingBase unreal = new TheUnrealThing();

Console.WriteLine(real has UnrealAttribute);
Console.WriteLine(unreal has UnrealAttribute);

```

prints:

```
False
True
```

Reasoning:

Would make using attributes MUCH easier (instead of using reflection), also this could probably be optimized in the compiler so the `has` keyword doesn't need to use reflection but use some form of fast access metadata instead.

Also improves readability.

Usage scenarios:

I would LOVE to use this in my automatic "read environment variables and parse them into objects" framework, also this would be super helpful for me when it comes to processing arbitrary data "channeled" from cloud to client.

**Comment by svick** [comment, marked as answer]
You're writing specialized code, so I think it's fine if you have to use some helper methods to make it easier. It could look like this:

```c#
public static bool Has<TAttribute>(this object obj) where TAttribute : Attribute =>
    obj.GetType().GetCustomAttribute<TAttribute>(inherit: false) is not null;

Console.WriteLine(real.Has<UnrealAttribute>());
Console.WriteLine(unreal.Has<UnrealAttribute>());
```

I think this is not much worse than the proposed syntax.

Regarding performance, I don't see any reason why the proposed syntax would be faster than reflection. If reflection is not fast enough, then that's what should be improved.

**Reply by bernd5** [reply]
The full code would be: [link](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwDoAlAUwDMAbegYwSgHsYBuY4gGUuYeggAWsAOYAhCAGd65ZPQjNyAXnIx6Ad3IAVcfQZqj0gBQBKfkWGiJ0uYvLoVazdr2HjAVRjuzOYwUta2xFQAnBaB1AASCgA8/oEAgggIyFDAqAj0AHzWNhF40W6qzPFJKRXpmdm5BUXhRCUAbJQ45AnydcgAogAeeTDy3DAA3sTkM5QAzJR4HcBcXOo9iQZ9DXmFjvLkXMAAVuwIhydW5LrGKobbOXnkIOQPjZr507PfR8fUAOJiAwATwADvRrACxABhVDyBAiN55TZIpqwW5QJDkRhqRRXKAHGBcc4wVDMZi8b7EAC+AhICwgwHhyAgHE65HsYkkIWc9GIUyIMwA9ELruIIHkAG70ZA0ulYBaKNT0AAm7KMJgqwSkzw5Ii5TgUfKIAuFopuEvo0tlRFprXp7JqalRutREztxAA2k7mABdCKKiqq9V+AJa7k6l6cxw8o38r4isWW61y+1AA=)

An optimization would be possible if the runtime type is known at compile type. In this case the expression could be replaced with a constant value (true / false).
In general this is not the case...

**Reply by jnm2** [reply]
That optimization is traditionally a kind of thing the JIT would see and perform too, because other steps like inlining would affect the ability to do the optimization.
